### ISSUE TITLE
too many sockets open when running ``app.run`` inside a loop

### ISSUE THREAD

**Issue by Dutchtallman** [pypr/pysph#247]
When calling app.run inside an optimization loop, python3 ends with an error:

```
  OSError: Unable to create file (unable to open file: name = '*.hdf5',
  errno = 24, error message = 'Too many open files', flags = 13, o_flags = 242)

```
When no output files are written, using the flag ``--disable-output``, then the
same error occurs, but now when intenting to write another file.

The modified example script ``bouncing_cube-modified.py`` reproduces the error.
Here the ``app.run`` is called from within a simple loop. At my system it runs
until the system soft limit of 512 files is hit. The scipt adds information as to which files are open for running the script.

I cannot find the root cause of this error, help needed here.

The error can be by-passed using 2 python scripts: one with the loop, calling the
other one using a ``subprocess``. 

, see ``loop.py`` running together with ``bouncing-cube2.py``

The files can be found here, rename the .txt files to .py

[bouncing_cube2.txt](https://github.com/pypr/pysph/files/4232512/bouncing_cube2.txt)
[bouncing_cube-modified.txt](https://github.com/pypr/pysph/files/4232513/bouncing_cube-modified.txt)
[loop.txt](https://github.com/pypr/pysph/files/4232514/loop.txt)

**Comment by prabhuramachandran** [comment]
Thanks for this and the reproducible scripts.  The scripts don't work on my mac since your file descriptor code is Linux specific.  In any case I am able to reproduce the issue with:
```
ulimit -n 32
python bouncing_cube-modified.py
```
which fails after 10 invocations.  

I see the problem, it is the multi-processing interface that we have.  Try this:

```
python bouncing_cube-modified.py --no-multiproc
```
That should work, I will see how we can fix this.

**Comment by prabhuramachandran** [comment]
I've fixed this in #259.
